virttest/vt_utils/block.py

```diff
@@ -0,0 +1,204 @@
+    start = utils_numeric.normalize_data_size(start, order_magnitude="M") + "M"
+    end = str(float(start[:-1]) + float(size[:-1])) + size[-1]
+    orig_disks = _list_disk_partitions()
+    partprobe_cmd = "partprobe /dev/%s" % did
+    mkpart_cmd = 'parted -s "%s" mkpart %s %s %s'
+    mkpart_cmd %= ("/dev/%s" % did, part_type, start, end)
+    process.system(mkpart_cmd)
+    process.system(partprobe_cmd, timeout=timeout)
+    partition_created = wait.wait_for(lambda: _list_disk_partitions() - orig_disks,
+                                      step=0.5, timeout=30)
+    if not partition_created:
+        raise RuntimeError('Failed to create partition.')
+    kname = partition_created.pop().split('/')[-1]
+    return kname
+
+
+def get_partition_attrs(partition):
+    """
+    Get partition attributes.
+
+    :param partition: like /dev/sdb1.
+    :return: dict like {'start': '512B', 'end': '16106127359B',
+                       'size': '16106126848B', 'type': 'primary'}
+    """
+    pattern = r'(/dev/.*)p(\d+)' if "nvme" in partition else r'(/dev/.*)(\d+)'
+    dev_name, part_num = re.match(pattern, partition).groups()
+    parted_cmd = 'parted -s %s unit B print' % dev_name
+    pattern = re.compile(r'%s\s+(?P<start>\d+\w+)\s+(?P<end>\d+\w+)\s+'
+                         r'(?P<size>\d+\w+)\s+(?P<type>\w+)' % part_num)
+    return pattern.search(process.run(parted_cmd).stdout_text, re.M).groupdict()
+
+
+def resize_partition(partition, size):
+    """
+    Resize partition.
+    Note: not support gpt type resize for linux guest.
+
+    :param partition: partition that to be shrunk, like /dev/sdb1.
+    :param size: resize partition to size, unit is B.
+    """
+    mountpoint, fstype = filesystem.get_mpoint_fstype(partition)
+    if filesystem.is_mounted(partition, dst=mountpoint, fstype=fstype):
+        filesystem.umount(partition, mountpoint, fstype=fstype)
+
+    # FIXME: if nvme device need support this function.
+    dev_name, part_num = re.match(r'(/dev/.*)(\d+)', partition).groups()
+    parted_cmd = 'parted -s %s print' % dev_name
+    part_attrs = get_partition_attrs(partition)
+    start_size = int(utils_numeric.normalize_data_size(part_attrs['start'],
+                                                       'B').split('.')[0])
+    end_size = (int(utils_numeric.normalize_data_size(size, 'B').split('.')[0])
+                - start_size)
+    process.system(' '.join((parted_cmd, 'rm %s' % part_num)))
+    resizepart_cmd = ' '.join((parted_cmd, 'unit B mkpart {0} {1} {2}'))
+    process.system(resizepart_cmd.format(part_attrs['type'], start_size,
+                                         end_size))
+    process.system('partprobe %s' % dev_name)
+
+    if fstype == 'xfs':
+        process.system('xfs_repair -n %s' % partition)
+    elif fstype.startswith("ext"):
+        process.system('e2fsck -f %s' % partition)
+    else:
+        raise NotImplementedError("%s is not supported!" % fstype)
```

Comment:
I remembered that we talked about decoupling this part, do we want to do it in this patch? cc @YongxueHong

Reply:
> I remembered that we talked about decoupling this part, do we want to do it in this patch? cc @YongxueHong

Hi @nickzhq 
@zhencliu 's comment is reasonable, we could focus on the implementation related to the block, and the manipulation related to the filesystem could be removed.
Thanks.